pkg/scheduler/queue/queue.go

```diff
@@ -239,6 +266,8 @@ func (q *RequestQueue) dispatcherLoop() {
 		case <-q.stopRequested:
 			// Nothing much to do here - fall through to the stop logic below to see if we can stop immediately.
 			stopping = true
+		case <-q.observeInflightRequests:
+			q.processObserveInflightRequests()
```

Comment:
isn't the dispatcher loop what's supposed to dispatch queries? Can we move this directly in the `running` goroutine (which mean we can also remove `observeInflightRequests`)

Reply:
The dispatcher loop actually does _everything_ - it processes querier operations updates and notifications to `forgetDisconnectedQueriers` as well as in the `requestsSent` and `requestsCompleted` in the exact same manner as it handles these `observeInflightRequests`.
 
I had done something similar to your idea previously for the `requestsSent` and `requestsCompleted` notification channels - essentially just use atomics to observe and update from outside the dispatcherLoop instead of message passing. Charles wanted to avoid atomics if at all possible, though we didn't do any benchmarking on atomic vs. non-atomic under various conditions.

I am not necessarily convinced using atomics to observe on these metrics would be a hot path since the dispatcherLoop is already single-threaded, but it is nice and consistent with what we have already.